File: products-go/internal/persistence/api/product_repository.go
package repository

import (
	"github.com/nao99/spring-boot-3-vs-go/products-go/internal/domain/model"
	"gorm.io/gorm"
)

type ProductRepository interface {
	FindLast20Products() *[]model.Product
	FindById(id uint64) *model.Product
	Save(product *model.Product)
}

type ProductRepositoryImpl struct {
	Session *gorm.DB
}

func (repo *ProductRepositoryImpl) FindLast20Products() *[]model.Product {
	var products *[]model.Product

	repo.Session.Order("id desc").Limit(20).Find(&products)

	return products
}

func (repo *ProductRepositoryImpl) FindById(id uint64) *model.Product {
	product := model.Product{}
	repo.Session.Find(&product, id)

	return &product
}

func (repo *ProductRepositoryImpl) Save(product *model.Product) {
	repo.Session.Save(&product)
}

File: products-spring-boot/src/main/java/org/ndbs/spring/graalvm/product/persistence/api/ProductRepository.java
package org.ndbs.spring.graalvm.product.persistence.api;

import org.ndbs.spring.graalvm.product.domain.model.Product;
import org.springframework.data.jdbc.repository.query.Query;
import org.springframework.data.repository.CrudRepository;

import java.util.Set;

/**
 * ProductRepository interface <br>
 *
 * <br>
 *
 * @author Nikolai Osipov <nao99.dev@gmail.com>
 * @since  2022-12-11
 */
public interface ProductRepository extends CrudRepository<Product, Long> {
    @Query("SELECT * FROM products ORDER BY id DESC LIMIT 20;")
    Set<Product> findLast20();
}
